docs/source/federation/gateway.mdx

```diff
@@ -0,0 +1,227 @@
+
+### Customizing responses
+
+Let's say that whenever our implementing services return an operation result to the gateway, they include a `Server-Id` header in the response. The value of the header uniquely identifies the service in our graph.
+
+When the gateway then responds to a client, we want _its_ `Server-Id` header to include the identifier for _every_ service that contributed to the response. In this case, we can tell the gateway to aggregate the various server IDs into a single, comma-separated list.
+
+The flow for processing a single operation from a client application then looks like this:
+
+```mermaid
+sequenceDiagram;
+  Client app->>Gateway: Sends GraphQL operation
+  Gateway->>Gateway: Generates query plan for operation
+  loop For each operation in the query plan
+    Gateway->>Implementing Services: Sends the operation to the applicable service
+    Implementing Services->>Gateway: Responds with result and Server-Id header
+    Gateway->>Gateway: Adds the returned Server-Id to the shared context
+  end
+  Gateway->>Gateway: Adds all Server-Ids in the shared context to the response header
+  Gateway->>Client app: Sends operation response
+```
+
+To implement this flow, we can use the `didReceiveResponse` callback of the `RemoteGraphQLDataSource` class to inspect each implementing service's result as it comes in. We can add the `Server-Id` to the shared `context` in this callback, then pull the full list from the `context` when sending the final response to the client.
+
+<ExpansionPanel title="Expand example">
+
+```javascript
+const { ApolloServer } = require('apollo-server');
+const { ApolloGateway, RemoteGraphQLDataSource } = require('@apollo/gateway');
+
+class DataSourceWithServerId extends RemoteGraphQLDataSource {
+  // highlight-start
+  async didReceiveResponse({ response, request, context }) {
+    // Parse the Server-Id header and add it to the array on context
+    const serverId = response.http.headers.get('Server-Id');
+    if (serverId) {
+      context.serverIds.push(serverId);
+    }
+
+    // Return the response, even when unchanged.
+    return response;
+  }
+  // highlight-end
+}
+
+const gateway = new ApolloGateway({
+  serviceList: [
+    { name: 'products', url: 'http://localhost:4001' }
+    // other services
+  ],
+  // highlight-start
+  buildService({ url }) {
+    return new DataSourceWithServerId({ url });
+  }
+  // highlight-end
+});
+
+const server = new ApolloServer({
+  gateway,
+  subscriptions: false, // Must be disabled with the gateway; see above.
+  context() {
+    return { serverIds: [] };
+  },
+  plugins: [
+    // highlight-start
+    {
+      requestDidStart() {
+        return {
+          willSendResponse({ context, response }) {
+            // Append our final result to the outgoing response headers
+            response.http.headers.append(
```

Comment:
It might be sometimes desirable, but it's somewhat surprising to me that the existing code didn't use `set` rather than [`append`](https://developer.mozilla.org/en-US/docs/Web/API/Headers/append) since it feels like this could create a condition with a comma-delimited list of comma-delimited attributes if `Server-Id` had already been set by another plugin or in other code?

e.g., ending up with potentially:

```
Server-Id: server1,server2 server1,server5
```

Not asking you for changes on this, but @trevor-scheer for input/thoughts.

Reply:
@abernix that's right, I don't see any reason why this shouldn't be `set` instead.